Clarifying gcr vs mcr pause image usage for Windows (#29063)
kubernetes/website: content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md

#### Pause Image

> Kubernetes maintains a multi-architecture image that includes support for Windows at `k8s.gcr.io/pause:3.5`.

Reviewer: ```suggestion
Kubernetes maintains a multi-architecture image as `k8s.gcr.io/pause`. From
version 3.5 (image tag ``k8s.gcr.io/pause:3.5`), that image includes support for Windows.
```
?

Author: I do want to be explicit about the version we are asking people to use to make sure the pause image contains images for each Windows OS version listed at https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#windows-os-version-support
 
In the past we've seen a lot of users run into issues where they were using an old pause image on newer OS releases so I would like to just give very prescriptive guidance here.

Author: @jsturtevant do you have any thoughts here?

Reviewer: How about “For Kubernetes v1.22, the recommended `pause` image is xxxx”. Then, every release, track a task to make sure to review that the advice is current.

(If you don't want that work every release, then instead I recommend a fuzzier wording that leaves the onus on the reader to check. What I think none of is want is a page that isn't maintained and ends up being outright inaccurate.)

Author: That works!
We are committed to updating the supporting Windows OS versions every release and I see making sure the docs specify an appropriate pause image that works with all of the supported Windows OS versions as part of that support.

> Source for the pause image can be found  [here](https://github.com/kubernetes/kubernetes/tree/master/build/pause).

Reviewer 2: I am not sure the best way to phrase this but we should avoid linking on the word `here` to better support screen readers.

Author: I had the same thought. Let me try and re-word this area.

> Source for the pause image can be found  [here](https://github.com/kubernetes/kubernetes/tree/master/build/pause).

Reviewer: ```suggestion
The [source code](https://github.com/kubernetes/kubernetes/tree/master/build/pause)
is available on GitHub.
```

> Microsoft maintains a multi-architecture image with Linux and Windows amd64 support at `mcr.microsoft.com/oss/kubernetes/pause:3.5`.

Reviewer: Similar to the earlier change, I'm wary of specifying a particular version. Is it OK that readers should use either the 3.5 version or any later version?

Author: Same comments as above.

#### Pause Image

Kubernetes maintains a multi-architecture image that includes support for Windows at `k8s.gcr.io/pause:3.5`.
The [source code](https://github.com/kubernetes/kubernetes/tree/master/build/pause)
is available on GitHub.

Microsoft maintains a multi-architecture image with Linux and Windows amd64 support at `mcr.microsoft.com/oss/kubernetes/pause:3.5`.
This image is built from the same source as the Kubernetes maintained image but all of the Windows binaries are [authenticode signed](https://docs.microsoft.com/en-us/windows-hardware/drivers/install/authenticode) by Microsoft.

> For this reason the Microsoft maintained image is recommended for production environments.

Reviewer 2: ```suggestion
The Microsoft maintained image is recommended for production environments when signed binaries are required
```

using "this" was a little confusing.

Author: Can you add the trailing . into you suggestion?
If you do that I'll commit the changes.

Reviewer 2: done

#### Pause Image

Kubernetes maintains a multi-architecture image `k8s.gcr.io/pause:3.5` that

> supports Linux as well as Windows.

Reviewer 3: "supports both Windows and Linux", the current wording is odd if you're no longer contrasting it with a Windows-only container image.

#### Pause Image

Kubernetes maintains a multi-architecture image `k8s.gcr.io/pause:3.5` that
supports Linux as well as Windows.

Microsoft maintains a Windows pause infrastructure container at
`mcr.microsoft.com/oss/kubernetes/pause:3.5`.

> This image is built from the same source as the Kubernetes maintained image but all of the Windows binaries are (authenticode signed)[https://docs.microsoft.com/en-us/windows-hardware/drivers/install/authenticode] by Microsoft.

Reviewer 2: We also don't build other archs such as `"architecture": "s390x",`

Author: I updated this to say the mcr image only container linux/windows amd64 images PTAL

Reviewer 4: The link syntax is wrong, it should be `[authenticode signed](https://blahblah)`.

Author: whoops!

* `kubectl port-forward` fails with "unable to do port forwarding: wincat not found"

  Port forwarding support for Windows was added in Kubernetes 1.15 by including wincat.exe in the

>   [pause infrastructure container](#Pause-Image).

Reviewer: We can omit this; Kubernetes versions before 1.15 are out of support.

Author: Updated/removed 1.15 reference.

* `kubectl port-forward` fails with "unable to do port forwarding: wincat not found"

  This was implemented in Kubernetes 1.15 by including wincat.exe in the

>   pause infrastructure container `k8s.gcr.io/pause:3.5`.

Reviewer 3: This change is a little weird... I read this as saying the version of the pause image that _first_ had wincat.exe, so wouldn't have expected it to update with new releases. Perhaps it shouldn't be the full container repository name, but just mention the appropriate version? (Although `k8s.gcr.io/pause` has _always_ had Wincat, as long as it's supported Windows, i.e. 3.4, I assume the previous, distinct Windows pause image versions (1.x) are where that was first introduced.)

Reviewer 3: wincat was introduced in `mcr.microsoft.com/k8s/core/pause:1.2.0` ([reference](https://github.com/kubernetes/website/pull/15345/files)), but that line's been updated several times since, so I guess "Be sure to use these versions or newer ones." hasn't been a strict requirement for a while, and can probably be removed.

Perhaps it should not bother naming the container image version which is adequately described elsewhere, and just say
> This was implemented in Kubernetes 1.15 by including wincat.exe in the pause infrastructure container.

since the following paragraph explains what to do if you're _not_ using a build of the k8s-specified pause image.

Author: Yea, i'll make these updates.
IIRC this did not specify a specific version of the pause image and then there was feedback saying we should include a version of the pause image for clarity. It has been so long now that all of the pause images folks can find should have wincat.exe (which hasn't changed) so not it makes sense to just drop the specific reference.

Author: I tried to re-word this. Let me know if it is more clear (and thanks for the feedback!)

Reviewer 3: Yeah, looks good and much clearer than before you started. Thank you.

  Be sure to use these versions or newer ones.  If you would like to build your
  own pause infrastructure container be sure to include

>   [wincat](https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/cmd/wincat).

Reviewer 3: This isn't the source for wincat in `k8s.gcr.io/pause:3.5`. It comes from https://github.com/kubernetes/kubernetes/tree/master/build/pause/windows/wincat now.

* `kubectl port-forward` fails with "unable to do port forwarding: wincat not found"

>   Port forwarding support for Windows was requires wincat.exe to be available in the

Reviewer 2: ```suggestion
  Port forwarding support for Windows requires wincat.exe to be available in the
```

  [pause infrastructure container](#pause-image).

>   Ensure you are using a supported image that is compatable with your Windows OS version.

Reviewer 2: Should we tell them how or what images are supported per version?

Reviewer 2: maybe as part of the section above where we give the specific version

Author: I like @sftim's suggestion to specify what the current recommended pause image is and just update that each release.
All of the past docs for the past releases recommend an image that will work for that release.

  to accommodate worker containers crashing or restarting without losing any of
  the networking configuration.

>   The "pause" (infrastructure) image can be found at `k8s.gcr.io/pause:3.5`

Reviewer 2: ```suggestion
  The "pause" (infrastructure) image can be found at `k8s.gcr.io/pause:3.5`.
```

  to accommodate worker containers crashing or restarting without losing any of
  the networking configuration.

>   The latest version of the pause image can be found above [here](#Pause-Image).

Reviewer 2: ```suggestion
Refer to the [pause image section](#Pause-Image) to find the  latest version of the pause image. ```

Author: Updated.

>   Refer to the [pause image section](#Pause-Image) to find the latest version of the pause image.

Reviewer: ```suggestion
  Refer to the [pause image](#pause-image) section to find the recommended version
  of the pause image.
```
(fragment identifiers are case sensitive in conformant browsers)

Author: Ah, I was following the case in in the rendered markdown https://github.com/kubernetes/website/blob/main/content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md#pause-image but when I copy/paste the link it does show up in all lowercase.

TIL :)

[here](/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/). We are
also making investments in cluster API to ensure Windows nodes are properly
provisioned.

> 

Reviewer 2: surprise it wasn't flagged for not having a blank line at end...

Author: Not sure how this got dropped but I added it back.
